File: integrations/typescript/posthog/src/PosthogSink.ts
import type { Sink, WideEventPayload } from '@axora/logger-core';

export interface PosthogClientLike {
  capture(payload: {
    distinctId: string;
    event: string;
    timestamp: Date;
    properties: Record<string, unknown>;
  }): void;
  shutdown(): Promise<void> | void;
}

export interface PosthogSinkOptions {
  client: PosthogClientLike;
}

export class PosthogSink implements Sink {
  constructor(private readonly options: PosthogSinkOptions) {}

  async export(event: WideEventPayload): Promise<void> {
    const distinctId = (event.context.user_id as string | undefined) ?? `service:${event.service}`;

    this.options.client.capture({
      distinctId,
      event: event.operation,
      timestamp: new Date(event.timestamp_start),
      properties: {
        ...event.context,
        axora_event_id: event.event_id,
        axora_service: event.service,
        status: event.status,
        level: event.level,
        duration_ms: event.duration_ms,
        ...(event.error.message ? { error_message: event.error.message } : {}),
      },
    });
  }

  async flush(): Promise<void> {
    await this.options.client.shutdown();
  }
}

File: integrations/csharp/posthog/PosthogSink.cs
using Axora.Logger.Sinks;

namespace Axora.Logger.Posthog;

public interface IPosthogClientLike
{
    void Capture(string distinctId, string eventName, IDictionary<string, object?> properties, DateTimeOffset timestamp);
    void Shutdown();
}

public sealed class PosthogSink : ISink
{
    private readonly IPosthogClientLike _client;

    public PosthogSink(IPosthogClientLike client)
    {
        _client = client;
    }

    public Task ExportAsync(WideEventPayload @event)
    {
        var distinctId = @event.Context.TryGetValue("user_id", out var userId) && userId is not null
            ? userId.ToString()!
            : $"service:{@event.Service}";

        var properties = new Dictionary<string, object?>(@event.Context)
        {
            ["axora_event_id"] = @event.EventId,
            ["axora_service"] = @event.Service,
            ["status"] = @event.Status,
            ["level"] = @event.Level,
            ["duration_ms"] = @event.DurationMs,
        };

        if (@event.Error.Message is not null)
        {
            properties["error_message"] = @event.Error.Message;
        }

        _client.Capture(distinctId, @event.Operation, properties, DateTimeOffset.Parse(@event.TimestampStart));
        return Task.CompletedTask;
    }

    public Task FlushAsync()
    {
        _client.Shutdown();
        return Task.CompletedTask;
    }
}
